pkg/cliplugins/workspace/plugin/kubeconfig.go

```diff
@@ -673,3 +675,109 @@ func newKCPClusterClient(clientConfig clientcmd.ClientConfig) (kcpclient.Cluster
 	clusterConfig.UserAgent = rest.DefaultKubernetesUserAgent()
 	return kcpclient.NewClusterForConfig(clusterConfig)
 }
+
+// ShowWorkspaceTreeOptions contains options for displaying the workspace tree
+type ShowWorkspaceTreeOptions struct {
+	*base.Options
+
+	kcpClusterClient kcpclient.ClusterInterface
+}
+
+// NewShowWorkspaceTreeOptions returns a new ShowWorkspaceTreeOptions.
+func NewShowWorkspaceTreeOptions(streams genericclioptions.IOStreams) *ShowWorkspaceTreeOptions {
+	return &ShowWorkspaceTreeOptions{
+		Options: base.NewOptions(streams),
+	}
+}
+
+// BindFlags binds fields to cmd's flagset.
+func (o *ShowWorkspaceTreeOptions) BindFlags(cmd *cobra.Command) {}
+
+// Complete ensures all dynamically populated fields are initialized.
+func (o *ShowWorkspaceTreeOptions) Complete() error {
+	if err := o.Options.Complete(); err != nil {
+		return err
+	}
+
+	kcpClusterClient, err := newKCPClusterClient(o.ClientConfig)
+	if err != nil {
+		return err
+	}
+	o.kcpClusterClient = kcpClusterClient
+
+	return nil
+}
+
+// Run outputs the current workspace.
+func (o *ShowWorkspaceTreeOptions) Run(ctx context.Context) error {
+	config, err := o.ClientConfig.ClientConfig()
+	if err != nil {
+		return err
+	}
+	_, currentClusterName, err := pluginhelpers.ParseClusterURL(config.Host)
+	if err != nil {
+		return fmt.Errorf("current URL %q does not point to cluster workspace", config.Host)
+	}
+
+	tree := treeprint.New()
+	b := tree.AddBranch("workspaces")
+	var result *v1beta1.WorkspaceList
```

Comment:
nit: you should be able to factor this into one caller, and not need to duplicate the list+recurse logic